blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+Let's have a look in the [official sources](https://github.com/ngrx/platform/blob/master/modules/store/src/selector.ts) how memoization works in the `createSelector` function. Every selector is created with `createSelectorFactory` and the argument for memoization, by default using `defaultMemoize`, which returns a `MemoizedProjection` for every selector doing the following:
+
+1. It checks if the incoming arguments have changed. If they didn't change (and this is not the first call), the memoized result is returned.
+2. If the arguments changed (or it is the first call), the projector function is applied.
+3. If the result didn't change, the memoized result is returned. If not, the new result is memoized and returned.
+
+This all might sound pretty straight forward, but pay attention to the fact, that the check for change is by default an object identity check and not a check for equality. With this, a selector might fire more often than you think.:
+
+```ts
+export function isEqualCheck(a: any, b: any): boolean {
+  return a === b;
+}
+```
+
+When updating a slice of state managed by `@ngrx/entity` the first check for input identity will always detect a change in selectors using that slice of state. So the projector function is most often applied and only the result identity check can then prevent the selector from firing.
+
+If we check our example selectors we can see how we are doing:
+
+- `getBook`:
+  
+  The projector of this selector references a single value of the entities state of `@ngrx/entity`. Updates to the state are handled by entity adapters and the object referenced will only change if the book with the same ID selected is updated or deleted. Other book updates will modify a different reference, so we are safe here. The third check in `defaultMemoize` will detect object identity for updates to other objects and thus the selector is properly memoized.
+
+- `getAuthorsOfBook`:
+
+  The fatal flaw lies within the projector of this selector:
+
+  ```ts
+  book ? book.authorIds.map(authorId => authors[authorId])` : []
+  ```
+  
+  This is a function that will always create a new array upon every call. So the third check for output identity will most certainly always fail and the selector will fire every time its input changes. In this case every change to the authors state - be it related to the currently selected book or not - will trigger a change.
```

Comment:
use longer dashes here
`–`

Reply:
ehem, how? https://stackoverflow.com/questions/52371734/how-to-make-a-long-in-markdown-i-e-the-character

Reply:
I'd just insert the character as is. Here's a few you can copy 😅

– – – – – – – – – –

Reply:
oh, that's plenty. I only needed 5 :sweat_smile: